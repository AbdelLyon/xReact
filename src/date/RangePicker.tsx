"use client";

import * as React from "react";
import { Calendar as CalendarIcon } from "lucide-react";
import { DateRange } from "react-day-picker";

import { cn } from "@/utils";
import { Popover, PopoverContent, PopoverTrigger } from "@/ui/popover";
import { Button } from "@/ui/button";
import { Calendar } from "@/ui/calendar";

type RangePickerProps = {
  className?: string;
  classNames?: {
    calendarIcon?: string;
    popoverContent?: string;
  };
  date?: DateRange;
  setDate: (date: DateRange | undefined) => void;
} & React.HTMLAttributes<HTMLDivElement>;

export const RangePicker = ({
  className,
  date,
  setDate,
  classNames,
  ...props
}: RangePickerProps) => {
  const formatDateRange = () => {
    if (!date?.from) return "Date de d√©but - Date de fin";

    return date.to
      ? `${date.from.toLocaleDateString()} - ${date.to.toLocaleDateString()}`
      : date.from.toLocaleDateString();
  };

  return (
    <div className={cn("grid gap-2", className)} {...props}>
      <Popover>
        <PopoverTrigger asChild>
          <Button
            aria-label="Select date range"
            className={cn("justify-start text-left font-normal dark:bg-card")}
            id="date"
            variant="outline"
          >
            <CalendarIcon
              className={cn("mr-2 size-4", classNames?.calendarIcon)}
            />
            {formatDateRange()}
          </Button>
        </PopoverTrigger>
        <PopoverContent
          align="start"
          className={cn("w-auto p-0", classNames?.popoverContent)}
        >
          <Calendar
            initialFocus
            defaultMonth={date?.from}
            mode="range"
            numberOfMonths={2}
            selected={date}
            onSelect={setDate}
          />
        </PopoverContent>
      </Popover>
    </div>
  );
};
