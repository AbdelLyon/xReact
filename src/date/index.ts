export { RangePicker } from "@/date/RangePicker";
export type { DateRange } from "react-day-picker";
