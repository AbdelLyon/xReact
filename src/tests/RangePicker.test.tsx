import { render } from "@testing-library/react";
import { describe, it, expect } from "vitest";
import { DateRange } from "react-day-picker";

import { RangePicker } from "@/date";

describe("RangePicker", () => {
  it("should match the snapshot with default props", () => {
    const { asFragment } = render(<RangePicker setDate={() => {}} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("should match the snapshot with selected date range", () => {
    const dateRange: DateRange = {
      from: new Date(2023, 8, 1),
      to: new Date(2023, 8, 15),
    };
    const { asFragment } = render(
      <RangePicker date={dateRange} setDate={() => {}} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should match the snapshot with custom class names", () => {
    const dateRange: DateRange = {
      from: new Date(2023, 8, 1),
      to: new Date(2023, 8, 15),
    };
    const { asFragment } = render(
      <RangePicker
        classNames={{
          calendarIcon: "text-red-500",
          popoverContent: "bg-gray-100",
        }}
        date={dateRange}
        setDate={() => {}}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
