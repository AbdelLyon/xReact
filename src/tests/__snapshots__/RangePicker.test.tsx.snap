// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RangePicker > should match the snapshot with custom class names 1`] = `
<DocumentFragment>
  <div
    class="grid gap-2"
  >
    <button
      aria-controls="radix-:r2:"
      aria-expanded="false"
      aria-haspopup="dialog"
      aria-label="Select date range"
      class="inline-flex items-center whitespace-nowrap rounded-md text-sm ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 justify-start text-left font-normal dark:bg-card"
      data-state="closed"
      id="date"
      type="button"
    >
      <svg
        class="lucide lucide-calendar mr-2 size-4 text-red-500"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 2v4"
        />
        <path
          d="M16 2v4"
        />
        <rect
          height="18"
          rx="2"
          width="18"
          x="3"
          y="4"
        />
        <path
          d="M3 10h18"
        />
      </svg>
      9/1/2023 - 9/15/2023
    </button>
  </div>
</DocumentFragment>
`;

exports[`RangePicker > should match the snapshot with default props 1`] = `
<DocumentFragment>
  <div
    class="grid gap-2"
  >
    <button
      aria-controls="radix-:r0:"
      aria-expanded="false"
      aria-haspopup="dialog"
      aria-label="Select date range"
      class="inline-flex items-center whitespace-nowrap rounded-md text-sm ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 justify-start text-left font-normal dark:bg-card"
      data-state="closed"
      id="date"
      type="button"
    >
      <svg
        class="lucide lucide-calendar mr-2 size-4"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 2v4"
        />
        <path
          d="M16 2v4"
        />
        <rect
          height="18"
          rx="2"
          width="18"
          x="3"
          y="4"
        />
        <path
          d="M3 10h18"
        />
      </svg>
      Date de d√©but - Date de fin
    </button>
  </div>
</DocumentFragment>
`;

exports[`RangePicker > should match the snapshot with selected date range 1`] = `
<DocumentFragment>
  <div
    class="grid gap-2"
  >
    <button
      aria-controls="radix-:r1:"
      aria-expanded="false"
      aria-haspopup="dialog"
      aria-label="Select date range"
      class="inline-flex items-center whitespace-nowrap rounded-md text-sm ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 justify-start text-left font-normal dark:bg-card"
      data-state="closed"
      id="date"
      type="button"
    >
      <svg
        class="lucide lucide-calendar mr-2 size-4"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 2v4"
        />
        <path
          d="M16 2v4"
        />
        <rect
          height="18"
          rx="2"
          width="18"
          x="3"
          y="4"
        />
        <path
          d="M3 10h18"
        />
      </svg>
      9/1/2023 - 9/15/2023
    </button>
  </div>
</DocumentFragment>
`;
